storage/include/ecs/s3_client.hpp

```diff
@@ -0,0 +1,393 @@
+    }
+  }
+
+  virtual Status object_exists_internal(const Sliver& key) const {
+    assert(init_);
+    ResponseData rData;
+    S3ResponseHandler rHandler;
+    rHandler.completeCallback = responseCompleteCallback;
+    rHandler.propertiesCallback = propertiesCallback;
+
+    S3_head_object(&context_, string(key.data()).c_str(), NULL, &rHandler, &rData);
+    LOG_DEBUG(
+        logger_,
+        "object_exist key: " << string(key.data()) << ", status: " << rData.status << ", msg: " << rData.errorMessage);
+    if (rData.status == S3Status::S3StatusOK)
+      return Status::OK();
+    else {
+      LOG_ERROR(logger_,
+                "object_exist key: " << string(key.data()) << ", status: " << rData.status
+                                     << ", msg: " << rData.errorMessage);
+      if (rData.status == S3Status::S3StatusHttpErrorNotFound || rData.status == S3Status::S3StatusErrorNoSuchBucket ||
+          rData.status == S3Status::S3StatusErrorNoSuchKey)
+        return Status::NotFound("Status: " + to_string(rData.status) + "msg: " + rData.errorMessage);
+
+      return Status::GeneralError("Status: " + to_string(rData.status) + "msg: " + rData.errorMessage);
+    }
+  }
+
+  virtual Status test_bucket_internal() {
+    assert(init_);
+    ResponseData rData;
+    S3ResponseHandler rHandler;
+    rHandler.completeCallback = responseCompleteCallback;
+    rHandler.propertiesCallback = propertiesCallback;
+    char location[100];
+
+    S3_test_bucket(context_.protocol,
+                   context_.uriStyle,
+                   context_.accessKeyId,
+                   context_.secretAccessKey,
+                   context_.hostName,
+                   context_.bucketName,
+                   100,
+                   location,
+                   NULL,
+                   &rHandler,
+                   &rData);
+    LOG_DEBUG(logger_,
+              "test_bucket bucket: " << context_.bucketName << ", status: " << rData.status
+                                     << ", msg: " << rData.errorMessage);
+    if (rData.status == S3Status::S3StatusOK)
+      return Status::OK();
+    else {
+      LOG_ERROR(logger_,
+                "test_bucket bucket: " << context_.bucketName << ", status: " << rData.status
+                                       << ", msg: " << rData.errorMessage);
+      if (rData.status == S3Status::S3StatusHttpErrorNotFound || rData.status == S3Status::S3StatusErrorNoSuchBucket)
+        return Status::NotFound("Status: " + to_string(rData.status) + "msg: " + rData.errorMessage);
+
+      return Status::GeneralError("Status: " + to_string(rData.status) + "msg: " + rData.errorMessage);
+    }
+  }
+
+  struct ResponseData {
+    S3Status status = S3Status::S3StatusOK;
+    string errorMessage;
+  };
+
+  /**
+   * @brief Represents intermidiate data for get operation and is passed to
+   * the lambda callback.
+   *
+   */
+  struct GetObjectResponseData : public ResponseData {
```

Comment:
Can we use an `std::vector` or an `std::string` (utilising `reseve()`) for simplicity?